src/instructlab/sdg/utils/chunkers.py

```diff
@@ -0,0 +1,609 @@
+                "meta-data",
+                "figure",
+                "page-header",
+            ]:
+                continue
+            if book_element["type"] == "footnote":
+                current_book_page_number = book_element["prov"][0]["page"]
+            elif book_element["type"] in [
+                "subtitle-level-1",
+                "paragraph",
+                "table",
+                "title",
+                "equation",
+            ]:  # 'page-header',
+                if book_element["type"] == "table":
+                    current_book_page_number = self.get_table_page_number(
+                        json_book, idx
+                    )
+                else:
+                    current_book_page_number = book_element["prov"][0]["page"]
+                    book_text = book_element["text"]
+
+                if book_element["type"] == "subtitle-level-1":
+                    if book_title is None:
+                        book_title = book_text
+                        book_text = f"# Title: **{book_text}**"
+                    else:
+                        book_text = f"## **{book_text}**"
+
+                if book_element["type"] == "title":
+                    book_text = f"# **{book_text}**"
+                if book_element["type"] == "page-header":
+                    book_text = f"Page Header: **{book_text}**\n\n"
+
+                if chunking_criteria is not None:
+                    # custom break function that can be used to chunk document
+                    if chunking_criteria(book_text):
+                        document_chunks.append("\n\n".join(current_buffer))
+                        current_buffer = []
+                elif (
+                    prev_page_number is not None
+                    and prev_page_number != current_book_page_number
+                ) and keep_same_page_thing_together:
+                    document_chunks.append("\n\n".join(current_buffer))
+                    current_buffer = []
+                else:
+                    if (
+                        self.get_token_count("\n\n".join(current_buffer), tokenizer)
+                        >= max_token_per_chunk
+                        and len(current_buffer) > 1
+                    ):
+                        chunk_text = "\n\n".join(current_buffer[:-1])
+                        print(
+                            f"Current chunk size {self.get_token_count(chunk_text, tokenizer)} and max is {max_token_per_chunk}"
+                        )
+
+                        document_chunks.append("\n\n".join(current_buffer[:-1]))
+
+                        if (
+                            self.get_token_count(current_buffer[-1], tokenizer)
+                            >= max_token_per_chunk
+                        ):
+                            print(
+                                f"This is too big a document to be left in the current buffer {self.get_token_count(current_buffer[-1], tokenizer)}"
```

Comment:
Maybe this would be useful as a warning level logging event?  Even for that purpose, it is not clear to me that this is well written for consumption by an end user.  For example, it is saying "too big a document" but if I understand it correctly, the issue is not the size of the _document_ but rather the size of `book_element["text"]`, which isn't going to mean much to an end user.  So if we do have a warning level log it might wind up being something more like "The following text was dropped from {document_name} because it was too long to fit into a single context for synthetic data generation: {current_buffer[-1]}".  (Not sure where you get `document_name` from, however -- maybe it is in json_book somewhere?)